src/google/protobuf/compiler/csharp/csharp_enum_field.cc

```diff
@@ -184,13 +176,100 @@ void EnumFieldGenerator::WriteEquals(Writer* writer) {
   }
 }
 void EnumFieldGenerator::WriteToString(Writer* writer) {
+  writer->WriteLine("PrintField(\"$0$\", $1$, $2$_, writer);",
+                    descriptor_->name(), has_property_name, name());
+}
+
+EnumOneofFieldGenerator::EnumOneofFieldGenerator(const FieldDescriptor* descriptor,
+						 int fieldOrdinal)
+  : EnumFieldGenerator(descriptor, fieldOrdinal) {
+  oneof_name =
+    UnderscoresToCamelCase(descriptor->containing_oneof()->name(), false);
+  oneof_property_name =
+    UnderscoresToCamelCase(descriptor->containing_oneof()->name(), true);
+
+  has_property_name = oneof_name + "Case_ == " + oneof_property_name +
+    "OneofCase." + property_name();
+  
+}
+
+EnumOneofFieldGenerator::~EnumOneofFieldGenerator() {
+}
+
+void EnumOneofFieldGenerator::GenerateMembers(Writer* writer) {
+  AddDeprecatedFlag(writer);
   if (SupportFieldPresence(descriptor_->file())) {
-    writer->WriteLine("PrintField(\"$0$\", has$1$, $2$_, writer);",
-                      descriptor_->name(), property_name(), name());
-  } else {
-    writer->WriteLine("PrintField(\"$0$\", $1$_, writer);",
-                      descriptor_->name(), name());
+    writer->WriteLine("public bool Has$0$ {", property_name());
+    writer->WriteLine("  get { return $0$; }", has_property_name);
+    writer->WriteLine("}");
+  }
+  AddPublicMemberAttributes(writer);
+  writer->WriteLine("public $0$ $1$ {", type_name(), property_name());
+  writer->WriteLine("  get { return $0$ ? ($1$) $2$_ : $3$; }",
```

Comment:
What I meant is that you are adding suffix "_" to oneof_name on many places to obtain the the field name).  so there is a concept of 
oneof_field_name()  { return oneof_name + "_"; }    (the naming of the function is not so important).
If you have this function and one day you decide you need to change the naming convention of the class field (perhaps because you realize there is some naming conflict), you can do it in that function and you don't have to look for 20 places in the code where you are appending underscore.